File: 프로그래머스/1/42840. 모의고사/모의고사.java
import java.util.Arrays;
import java.util.ArrayList;

class Solution {
    public int[] solution(int[] answers) {
        
        // 초기화
        int[][] patterns = {
                {1,2,3,4,5},
                {2,1,2,3,2,4,2,5},
                {3,3,1,1,2,2,4,4,5,5}
        };
        int[] score = new int[3];
        ArrayList<Integer> answer = new ArrayList<>();
        
        
        // 수포자 답안과 정답 채점
        for(int i=0; i < answers.length; i++) {
            for(int j=0; j < patterns.length; j++) {
                if(answers[i] == patterns[j][i % patterns[j].length]) {
                    score[j]++;
                }
            }
        }
        
        int maxScore = Arrays.stream(score).max().getAsInt();
        for(int i=0; i < score.length; i++) {
            if(score[i] == maxScore)
                answer.add(i + 1);
        }
        
        return answer.stream().sorted().mapToInt(Integer::intValue).toArray();
    }
}

File: 프로그래머스/1/42840. 모의고사/모의고사.kt
class Solution {
    fun solution(answers: IntArray): IntArray {
        val result = mutableListOf<Int>()

        // 수포자 삼인방의 찍는 패턴
        val patterns = arrayOf(
            intArrayOf(1, 2, 3, 4, 5),
            intArrayOf(2, 1, 2, 3, 2, 4, 2, 5),
            intArrayOf(3, 3, 1, 1, 2, 2, 4, 4, 5, 5)
        )

        // 성적을 담을 배열
        val scores = IntArray(patterns.size)

        // 정답 채점
        for (i in answers.indices) {
            for (j in patterns.indices) {
                if (answers[i] == patterns[j][i % patterns[j].size]) {
                    scores[j]++
                }
            }
        }

        val maxScore = scores.maxOrNull()
        for (i in scores.indices) {
            if (scores[i] == maxScore) {
                result.add(i + 1)
            }
        }

        return result.toIntArray()
    }
}
